pkg/providers/imagefamily/agentbakerbootstrap/provisionclientbootstrap.go

```diff
@@ -0,0 +1,239 @@
+			}
+		} else {
+			provisionProfile.GpuProfile = &models.GPUProfile{
+				DriverType:       lo.ToPtr(models.DriverType_CUDA),
+				InstallGPUDriver: lo.ToPtr(true),
+			}
+		}
+
+	}
+
+	if p.IsWindows {
+		provisionProfile.OsType = lo.ToPtr(models.OSType_Windows)
+	} else {
+		provisionProfile.OsType = lo.ToPtr(models.OSType_Linux)
+	}
+
+	provisionHelperValues := &models.ProvisionHelperValues{
+		SkuCPU:    lo.ToPtr(p.InstanceType.Capacity.Cpu().AsApproximateFloat64()),
+		SkuMemory: lo.ToPtr(p.InstanceType.Capacity.Memory().AsApproximateFloat64()),
+	}
+
+	return p.getNodeBootstrappingFromClient(ctx, provisionProfile, provisionHelperValues, p.KubeletClientTLSBootstrapToken)
+}
+
+func (p *ProvisionClientBootstrap) getNodeBootstrappingFromClient(ctx context.Context, provisionProfile *models.ProvisionProfile, provisionHelperValues *models.ProvisionHelperValues, bootstrapToken string) (string, string, error) {
+	transport := httptransport.New(options.FromContext(ctx).NodeBootstrappingServerURL, "/", []string{"http"})
+	client := client.New(transport, strfmt.Default)
+
+	params := operations.NewNodeBootstrappingGetParams()
+	params.ResourceGroupName = p.ClusterResourceGroup
+	params.ResourceName = p.ClusterName
+	params.SubscriptionID = p.SubscriptionID
+	provisionValues := &models.ProvisionValues{
+		ProvisionProfile:      provisionProfile,
+		ProvisionHelperValues: provisionHelperValues,
+	}
+	params.Parameters = provisionValues
+
+	params.WithTimeout(30 * time.Second)
+	params.Context = ctx
+
+	resp, err := client.Operations.NodeBootstrappingGet(params)
+	if err != nil {
+		return "", "", err
+	}
+
+	if resp.Payload == nil {
+		return "", "", fmt.Errorf("no payload in response")
+	}
+	if resp.Payload.Cse == nil || *resp.Payload.Cse == "" {
+		return "", "", fmt.Errorf("no CSE in response")
+	}
+	if resp.Payload.CustomData == nil || *resp.Payload.CustomData == "" {
+		return "", "", fmt.Errorf("no CustomData in response")
+	}
+
+	cseWithoutBootstrapToken := *resp.Payload.Cse
+	customDataWithoutBootstrapToken := *resp.Payload.CustomData
+
+	cseWithBootstrapToken := strings.ReplaceAll(cseWithoutBootstrapToken, "{{.TokenID}}.{{.TokenSecret}}", bootstrapToken)
+
+	decodedCustomDataWithoutBootstrapTokenInBytes, err := base64.StdEncoding.DecodeString(customDataWithoutBootstrapToken)
+	if err != nil {
+		return "", "", err
+	}
+	decodedCustomDataWithBootstrapToken := strings.ReplaceAll(string(decodedCustomDataWithoutBootstrapTokenInBytes), "{{.TokenID}}.{{.TokenSecret}}", bootstrapToken)
+	customDataWithBootstrapToken := base64.StdEncoding.EncodeToString([]byte(decodedCustomDataWithBootstrapToken))
+
+	return customDataWithBootstrapToken, cseWithBootstrapToken, nil
+}
+
+func getAgentbakerGeneratedLabels(nodeResourceGroup string, nodeLabels map[string]string) {
+	// Delegatable defaulting?
+	nodeLabels["kubernetes.azure.com/role"] = "agent"
+	nodeLabels["kubernetes.azure.com/cluster"] = normalizeResourceGroupNameForLabel(nodeResourceGroup)
+}
```

Comment:
Do we need to make the karpenter nodepool aware of these defaulted labels?

Reply:
Maybe, though this would only be consequential if there are workloads targeting these, which is very unlikely. (More generally, it may be useful to develop a mechanism for letting Karpenter know Nodes will have certain labels - without having to specify them in NodePool.)

Reply:
FYI those are already there in self-contained.
The new one we will get from the provision server (and AgentBaker) will be the below. 
The emptiness is intentional to prevent workload targeting them, for now.
```
agentpool=
kubernetes.azure.com/agentpool=
```